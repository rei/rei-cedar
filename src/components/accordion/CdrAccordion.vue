<script>
import {
  defineComponent, useCssModule, computed, watch, onMounted, ref, inject,
} from 'vue';
import IconCaretDown from '../icon/comps/caret-down.vue';
import { modifyClassName } from '../../utils/buildClass';
import mapClasses from '../../utils/mapClasses';
/** Vertically-stacked list that allows users to expand and collapse additional content */
export default defineComponent({
  name: 'CdrAccordion',
  components: {
    IconCaretDown,
  },
  props: {
    /**
     * The unique id of an accordion.
     * @demoIgnore true
     */
    id: {
      type: String,
      required: true,
    },
    /**
     * Toggle this value to open/close the accordion.
     * @demoIgnore true
     */
    opened: {
      type: Boolean,
      default: false,
    },
    /**
     * Sets a compact style.
     */
    compact: {
      type: Boolean,
      default: false,
    },
    /**
     * Sets a border-aligned style.
     */
    borderAligned: {
      type: Boolean,
      default: false,
    },
    /**
     * Sets the heading level
     * @demoIgnore true
     */
    level: {
      type: [String, Number],
      required: true,
    },
    /**
     * Toggles content spacing (padding)
     */
    contentSpacing: {
      type: Boolean,
      default: true,
    },
    /**
     * Sets the readable text on CdrAccordion button (also can be slotted)
     * @demoIgnore true
     */
    label: {
      type: String,
    },
  },
  emits: {
    /**
    * Emits on accordion open/close
    */
    'accordion-toggle': null,
  },

  setup(props, ctx) {
    const unwrap = inject('unwrap', false);
    const isGrouped = inject('grouped', false) ? 'li' : 'div';
    const style = useCssModule();
    const accordionContentEl = ref(null);
    const focused = ref(false);
    const maxHeight = ref(props.opened
      ? 'none'
      : '0px');
    const headingTag = `h${props.level}`;
    const labelClass = 'cdr-accordion__label';
    const baseClass = 'cdr-accordion';
    const iconClass = 'cdr-accordion__icon';
    const containerClass = 'cdr-accordion__content-container';
    const contentClass = 'cdr-accordion__content';
    const isUnwrapped = computed(() => unwrap && unwrap.isUnwrapped);
    const headingContent = computed(() => (isUnwrapped.value
      ? 'div'
      : 'button'));
    const headingContentStyle = computed(() => (isUnwrapped.value
      ? 'js-cdr-accordion-button'
      : [style['cdr-accordion__button'], 'js-cdr-accordion-button']));
    const headingClass = computed(() => (isUnwrapped.value
      ? 'cdr-accordion__header--unwrapped'
      : 'cdr-accordion__header'));
    const compactClass = computed(() => (props.compact
      ? modifyClassName(baseClass, 'compact')
      : ''));
    const borderAlignedClass = computed(() => (props.borderAligned
      ? modifyClassName(baseClass, 'border-aligned')
      : ''));
    const focusedClass = computed(() => (focused.value
      ? modifyClassName(baseClass, 'focused')
      : null));
    const unwrapClass = computed(() => (isUnwrapped.value
      ? modifyClassName(baseClass, 'unwrap')
      : null));
    const noSpacingClass = computed(() => (!props.contentSpacing
      ? modifyClassName(baseClass, 'no-spacing')
      : null));

    const isOpenClass = computed(() => (props.opened || isUnwrapped.value
      ? 'cdr-accordion--open'
      : 'cdr-accordion--closed'));

    const onClick = (event) => {
      ctx.emit('accordion-toggle', event);
    };
    const onFocus = () => {
      focused.value = true;
    };
    const onBlur = () => {
      focused.value = false;
    };

    const listeners = computed(() => (isUnwrapped.value
      ? {}
      : {
        click: onClick,
        focus: onFocus,
        blur: onBlur,
      }));
    watch(() => props.opened, (opened) => {
      maxHeight.value = !opened ? `${accordionContentEl.value.clientHeight}px` : '0px';
      // nextTick is not sufficient here, must wait for CSS to re-paint
      setTimeout(() => {
        // on next frame, set maxHeight to new value
        maxHeight.value = opened ? `${accordionContentEl.value.clientHeight}px` : '0px';
        setTimeout(() => {
          // after animation is complete, remove max-height so content can reflow
          maxHeight.value = opened ? 'none' : '0px';
        }, 350); // cdr-duration-3x + 50ms
      }, 50);
    });

    onMounted(() => {
      if (props.opened && accordionContentEl.value) {
        maxHeight.value = 'none';
      }
    });
    return {
      style: useCssModule(),
      headingTag,
      headingClass,
      headingContent,
      headingContentStyle,
      isGrouped,
      unwrap,
      accordionContentEl,
      focused,
      maxHeight,
      baseClass,
      labelClass,
      compactClass,
      borderAlignedClass,
      focusedClass,
      iconClass,
      containerClass,
      isOpenClass,
      contentClass,
      noSpacingClass,
      listeners,
      unwrapClass,
      isUnwrapped,
      mapClasses,
    };
  },
});

</script>

<template>
  <component
    :is="isGrouped"
    :class="!isUnwrapped
      ? mapClasses(style, baseClass, compactClass, borderAlignedClass, focusedClass, noSpacingClass)
      : null"
    :id="`${id}-accordion`"
  >
    <component
      :is="headingTag"
      :class="style[headingClass]"
    >
      <!--
        Triggered on accordion open/close
        @event accordion-toggle
      -->
      <component
        :is="headingContent"
        :class="headingContentStyle"
        :id="id"
        v-on="listeners"
        :aria-expanded="!isUnwrapped ? `${opened}` : null"
        :aria-pressed="!isUnwrapped ? `${opened}` : null"
        :aria-controls="!isUnwrapped ? `${id}-collapsible` : null"
      >
        <span
          :class="style[labelClass]"
          :id="`${id}-label`"
        >
          <!-- @slot  Sets the readable text on the CdrAccordion button -->
          <slot name="label">
            {{ label }}
          </slot>
        </span>
        <icon-caret-down
          v-if="!isUnwrapped"
          :class="mapClasses(style, iconClass, isOpenClass)"
          :size="compact ? 'small' : null"
        />
      </component>
    </component>
    <div
      :class="mapClasses(style, containerClass, isOpenClass, unwrapClass)"
      :style="{ maxHeight: isUnwrapped ? 'none' : maxHeight }"
    >
      <div
        :class="mapClasses(style, contentClass, isOpenClass, unwrapClass)"
        :aria-hidden="!isUnwrapped ? `${!opened}` : null"
        :id="`${id}-collapsible`"
        ref="accordionContentEl"
      >
        <!-- @slot CdrAccordion content -->
        <slot />
      </div>
    </div>
  </component>
</template>

<style lang="scss" module src="./styles/CdrAccordion.module.scss">
</style>
