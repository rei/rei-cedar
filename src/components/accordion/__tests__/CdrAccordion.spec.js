import { shallowMount, mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrAccordion from '../CdrAccordion.vue';

const propsData = {
  id: 'test',
  label: 'A label',
  level: '2',
  compact: false,
  borderAligned: false,
};

describe('CdrAccordion', () => {
  it('renders correctly', () => {
    const wrapper = mount(CdrAccordion, {
      propsData: {
        id: 'test',
        level: '2',
      },
      slots: {
        default: 'This is some slot text.',
        label: 'label'
      },
    });
    expect(wrapper.element).toMatchSnapshot()
  });

  it('renders correctly unwrapped', async () => {
    const wrapper = mount(CdrAccordion, {
      propsData: {
        id: 'test',
        level: '2',
      },
      slots: {
        default: 'This is some slot text.',
        label: 'label'
      },
      global: {
        provide: {
          unwrap: {
            isUnwrapped: {
              value: true
            }
          }
        }
      }
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.element).toMatchSnapshot()
  });

  // TODO add more checks here
  it('meets a11y requirements', async () => {
    const wrapper = shallowMount(CdrAccordion, {
      propsData: {
        id: 'test',
        level: '2',
      },
      slots: {
        default: 'This is some slot text.',
        label: 'label',
      },
    });
    const button = wrapper.find('#test');
    const contentArea = wrapper.find('.cdr-accordion__content');
    expect(button.classes()).toContain('js-cdr-accordion-button');
    // closed state
    expect(button.attributes('aria-expanded')).toBe('false');
    expect(button.attributes('aria-controls')).toBe(`${wrapper.vm.id}-collapsible`);

    // opened state
    button.trigger('click');
    wrapper.setProps({ opened: true }); // fake the opening logic
    await wrapper.vm.$nextTick();
    expect(button.attributes('aria-expanded')).toBe('true');
    expect(contentArea.attributes('aria-hidden')).toBe('false');
  })

  describe('accordion data setup', () => {
    it('sets maxHeight when starting closed', () => {
      const wrapper = shallowMount(CdrAccordion, {
        propsData: {
          id: 'test',
          level: '2',
          opened: false,
        },
        slots: {
          default: 'This is some slot text.',
          label: 'label',
        },
      });

      expect(wrapper.vm.maxHeight).toBe(0);
    });

    it('maxHeight is adjusted in created when starting open', () => {
      const wrapper = shallowMount(CdrAccordion, {
        propsData: {
          id: 'test',
          level: '2',
          opened: true,
        },
        slots: {
          default: 'This is some slot text.',
          label: 'label',
        },
      });

      expect(wrapper.vm.maxHeight).toBe('none');
    });
  });

  describe('toggling accordion', () => {
    it('updates emits an event onClick', async () => {
      const wrapper = shallowMount(CdrAccordion, {
        propsData: {
          id: 'test',
          level: '2',
        },
        slots: {
          default: 'This is some slot text.',
          label: 'label',
        },
      });

      wrapper.find('button').trigger('click');
      await wrapper.vm.$nextTick();
      expect(wrapper.emitted('accordion-toggle'));
    });

    it('updates maxHeight on prop update', async () => {
      const wrapper = shallowMount(CdrAccordion, {
        propsData: {
          id: 'test',
          level: '2',
          opened: true,
        },
        slots: {
          default: 'This is some slot text.',
          label: 'label',
        },
      });

      wrapper.setProps({ opened: false });
      await wrapper.vm.$nextTick();
      expect(wrapper.vm.maxHeight).toBe('0px');
    });
  });

  it('isOpenClass computed prop', () => {
    const wrapper = shallowMount(CdrAccordion, {
      propsData: {
        id: 'test',
        level: '2',
        opened: false,
      },
      slots: {
        default: 'This is some slot text.',
        label: 'label',
      },
    });

    expect(wrapper.vm.isOpenClass).toEqual('cdr-accordion--closed');
  });

  it('focused style', async () => {
    const wrapper = shallowMount(CdrAccordion, {
      propsData: propsData,
      slots: {
        default: 'This is some slot text.',
        label: 'label',
      },
    });
    wrapper.find('button').trigger('focus');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.focused).toBeTruthy();
    expect(wrapper.classes()).toContain('cdr-accordion--focused');
    wrapper.find('button').trigger('blur');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.focused).toBeFalsy();
  });
  it('no-spacing style', async () => {
    const wrapper = shallowMount(CdrAccordion, {
      propsData: {
        id: 'test',
        level: '2',
        contentSpacing: false,
      },
      slots: {
        default: 'This is some slot text.',
        label: 'label',
      },
    });
    expect(wrapper.classes()).toContain('cdr-accordion--no-spacing');
  });
  it('style class checks prop values', () => {
    const wrapper = shallowMount(CdrAccordion, {
      propsData: {
        id: 'test',
        level: '2',
        compact: true,
        borderAligned: true,
      },
      slots: {
        default: 'This is some slot text.',
        label: 'label',
      },
    });

    expect(wrapper.classes()).toContain('cdr-accordion--compact');
    expect(wrapper.classes()).toContain('cdr-accordion--border-aligned');
  });
});
