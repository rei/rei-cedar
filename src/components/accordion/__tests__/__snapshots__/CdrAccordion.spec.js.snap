// Vitest Snapshot v1

exports[`CdrAccordion > component snapshot when unwrapped > renders correctly 1`] = `
<div
  class=""
  id="test-accordion"
>
  <h2
    class="cdr-accordion__header--unwrapped"
  >
    <!--
        Triggered on accordion open/close
        @event accordion-toggle
      -->
    <div
      class="js-cdr-accordion-button"
      id="test"
    >
      <span
        class="cdr-accordion__label"
        id="test-label"
      >
        <!-- @slot  Sets the readable text on the CdrAccordion button -->
        
        label
        
      </span>
      <!--v-if-->
    </div>
  </h2>
  <div
    class="cdr-accordion__content-container cdr-accordion--open cdr-accordion--unwrap"
    style="max-height: none;"
  >
    <div
      class="cdr-accordion__content cdr-accordion--open cdr-accordion--unwrap"
      id="test-collapsible"
    >
      <!-- @slot CdrAccordion content -->
      
      This is some slot text.
      
    </div>
  </div>
</div>
`;

exports[`CdrAccordion > component snapshot when wrapped > renders correctly 1`] = `
<div
  class="cdr-accordion"
  id="test-accordion"
>
  <h2
    class="cdr-accordion__header"
  >
    <!--
        Triggered on accordion open/close
        @event accordion-toggle
      -->
    <button
      aria-controls="test-collapsible"
      aria-expanded="false"
      aria-pressed="false"
      class="cdr-accordion__button js-cdr-accordion-button"
      id="test"
    >
      <span
        class="cdr-accordion__label"
        id="test-label"
      >
        <!-- @slot  Sets the readable text on the CdrAccordion button -->
        
        label
        
      </span>
      <svg
        aria-hidden="true"
        class="cdr-icon cdr-accordion__icon cdr-accordion--closed"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        
        
        
        <path
          d="M12 16c.273 0 .521-.11.702-.288l5.005-5.005a1 1 0 00-1.414-1.414L12 13.586 7.705 9.29a1 1 0 00-1.412 1.417l4.98 4.98c.182.193.44.313.727.313z"
          role="presentation"
        />
        
        <!--v-if-->
      </svg>
    </button>
  </h2>
  <div
    class="cdr-accordion__content-container cdr-accordion--closed"
    style="max-height: 0px;"
  >
    <div
      aria-hidden="true"
      class="cdr-accordion__content cdr-accordion--closed"
      id="test-collapsible"
    >
      <!-- @slot CdrAccordion content -->
      
      This is some slot text.
      
    </div>
  </div>
</div>
`;
